### ISSUE TITLE
DeAOTL predictions

### ISSUE THREAD

**Issue by bhack** [yoxu515/aot-benchmark#29]
Where are the DeAOTL predictions?

**Comment by bhack** [comment]
Also testing it at fullres is going to have some memory problem.

The first is the in memory accumulation with the separate saving stage design (but it is easy to workaround this):
https://github.com/yoxu515/aot-benchmark/blob/1c3a5ec51d81f3e17ff9092aa1e830206d766132/networks/managers/evaluator.py#L480****

But there is a second one as it seems that the memory is growing over the sequence and often it will go OOM before fully completing the sequence.

**Comment by bhack** [comment]
@jvntf I saw your diff at https://github.com/runwayml/aot-benchmark/commit/22cc7c59fc98fd73852c304737a6da65687f7197

Have you found the memory issue?

**Comment by z-x-yang** [comment]
How long is your sequence? What's the resolution? How much GPU memory do you have?

**Comment by bhack** [comment]
Around 100 FPS it stop near >70 on 16GB memory res 1920 × 1080.

**Comment by z-x-yang** [comment]
You could try to set --max_resolution to a smaller number (such as 1.3*480p in our default setting) to save memory.

**Comment by z-x-yang** [comment]
Since DeAOT-L/AOT-L utilizes the memory reading strategy proposed by STM, the memory will grow up during the inference of a sequence. To save memory, you could try to set [TEST_LONG_TERM_MEM_GAP](https://github.com/yoxu515/aot-benchmark/blob/1c3a5ec51d81f3e17ff9092aa1e830206d766132/configs/models/deaotl.py#L13) in the config file to a larger number (5 in default following STM) to reduce the frequency of long-term memory update.

Also, you could try to use DeAOT-T/S/B, which doesn't use the memory reading strategy and thus has consistent memory usage during inference.

**Comment by bhack** [comment]
This is interesting. Is there a long term mem gap optimal allocation based on the sequence length other then for memory saving?

**Comment by z-x-yang** [comment]
According to some of the experiments, if the sequence is 30fps, setting the gap to around 10~20 is a choice better than 5, leading to better results and less memory usage.

**Comment by bhack** [comment]
Increasing `TEST_LONG_TERM_MEM_GAP` helped in the memory consumption but still crashing later. 

What is the point about handling longer sequence e.g. 150/200 frames? Is the memory growing without a limit?

**Comment by z-x-yang** [comment]
We didn't set a limitation for memory growing, so you may encounter the OOM problem when the sequence is too long.

However, we have tried to use AOT-based methods to process videos with thousands of frames by modifying the current code, and won two champions in [the Visual Object Tracking (VOT) 2022 challenge](https://www.votchallenge.net/vot2022/trackers.html). The source code of our tracker can be downloaded from [here](http://data.votchallenge.net/vot2022/trackers/MS_AOT-code-2022-05-04T08_08_40.403273.zip).

In short, we set a maximum number of long-term memory frames, such as 10. If the total frame number of long-term memory is larger than the maximum number after updating the memory, we will drop the oldest frame (except for the first reference frame, which will be kept persistently) directly to make sure the long-term memory has a limited size.

Such a simple manner can solve the OOM problem and slightly improve the accuracy of tracking and segmentation.

**Comment by bhack** [comment]
Interesting, thanks. I will try to experiment with this strategy.

**Comment by bhack** [comment]
I am trying to just increase the resolution of the input and the memory It is going to OOM on the first frame (also with 40GB). 

Creating the global attention at `local2global` is already saturating the memory:

https://github.com/yoxu515/aot-benchmark/blob/1c3a5ec51d81f3e17ff9092aa1e830206d766132/networks/layers/attention.py#L390

**Comment by z-x-yang** [comment]
Can't help. Perhaps you could try smaller DeAOT models.

**Comment by bhack** [comment]
Yes I've already tried, I will try to figure out if there is any memory reduction opportunity in the code.

If any sections come to mind, let me know.

**Comment by bhack** [comment]
I was thinking, is there any major blocker to sliding window at inference/eval?

**Comment by bhack** [comment]
I meant something like:
https://docs.monai.io/en/1.1.0/inferers.html#sliding-window-inference
But I suppose that we could have some problem with the short/long memory managment.

**Comment by z-x-yang** [comment]
Thank you for your suggestion; however, the sliding-window inference may not be well-suited for VOS. Considering that objects within videos may continuously move or experience dramatic changes in size, it becomes difficult to identify an appropriate sliding window that consistently encompasses the object.

**Comment by z-x-yang** [comment]
I plan to close this issue. But if you still have some problems, please reopen it.
